ssm_cache/cache.py

```diff
@@ -5,16 +5,101 @@
 from functools import wraps
 import six
 
-import boto3
-
 class InvalidParameterError(Exception):
     """ Raised when something's wrong with the provided param name """
 
+class SSMParameterStore(object): # pylint: disable=too-few-public-methods
+    """Some hooks are provided to override behavior: see inject_session_factory and inject_client_factory
+    for details."""
+
+    @classmethod
+    def _default_session_factory(cls):
+        """Default session factory that creates a boto3 session."""
+        import boto3
+        return boto3.session.Session()
+
+    @classmethod
+    def _default_client_factory(cls, session, name):
+        """Default client factory that creates a client from the provided session."""
+        return session.client(name)
+
+    _SESSION_FACTORY = _default_session_factory
+    _CLIENT_FACTORY = _default_client_factory
+
+    _SESSION = None
+    _CLIENT = None
+
+    @classmethod
+    def inject_session_factory(cls, session_factory):
```

Comment:
Looking good!

I would only add a bit of validation. The given factory shouldn't be None and should be callable.

We could simply add this line:

```python
assert callable(session_factory), "Session factory must be callable"
```

Reply:
Do we really need this check? It's for advanced users who should know, and it will fail in an obvious way later.

Reply:
I just prefer when my code "explodes" asap, rather than downstream. So you can find what to fix in your code without necessarily learning/inspecting the library code.

Reply:
I've added the `assert` but am a bit worried about assertions getting disabled in the case of `-O` being used or `PYTHONOPTIMIZE` being set in the environment. Thoughts?

Reply:
That's a valid point. Asserts are better suited for test cases.

Raising a proper error will be as simple as:

```python
if not callable(session_factory):
    raise TypeError("Session factory must be callable")
```

Reply:
OK